be/src/olap/tablet.cpp

```diff
@@ -1446,4 +1437,99 @@ void Tablet::reset_compaction(CompactionType compaction_type) {
     }
 }
 
+Status Tablet::create_initial_rowset(const int64_t req_version) {
+    Status res = Status::OK();
+    if (req_version < 1) {
+        LOG(WARNING) << "init version of tablet should at least 1. req.ver=" << req_version;
+        return Status::OLAPInternalError(OLAP_ERR_CE_CMD_PARAMS_ERROR);
+    }
+    Version version(0, req_version);
+    RowsetSharedPtr new_rowset;
+    do {
+        // there is no data in init rowset, so overlapping info is unknown.
+        std::unique_ptr<RowsetWriter> rs_writer;
+        res = create_rowset_writer(version, VISIBLE, OVERLAP_UNKNOWN, &rs_writer);
+        if (!res.ok()) {
+            LOG(WARNING) << "failed to init rowset writer for tablet " << full_name();
+            break;
+        }
+        res = rs_writer->flush();
+        if (!res.ok()) {
+            LOG(WARNING) << "failed to flush rowset writer for tablet " << full_name();
+            break;
+        }
+
+        new_rowset = rs_writer->build();
+        res = add_rowset(new_rowset);
+        if (!res.ok()) {
+            LOG(WARNING) << "failed to add rowset for tablet " << full_name();
+            break;
+        }
+    } while (0);
+
+    // Unregister index and delete files(index and data) if failed
+    if (!res.ok()) {
+        LOG(WARNING) << "fail to create initial rowset. res=" << res << " version=" << req_version;
+        StorageEngine::instance()->add_unused_rowset(new_rowset);
+        return res;
+    }
+    set_cumulative_layer_point(req_version + 1);
+    return res;
+}
+
+Status Tablet::create_rowset_writer(const Version& version, const RowsetStatePB& rowset_state,
```

Comment:
It could not be one method, because one method is for load purpose so it have load id and txn id, and the other method is for existing rowset so it have version.
Although we could combine there 2 method to a single method, but has to add a boolean variable to indicate which variable is useful and user has to set default value during call this method.  I am not preferred with this way.

Reply:
But if a new member is added into RowsetWriterContext, you have to added in the two function, it's better to collect the same lines in a private function. Such as _init_context();

Reply:
I think it is affordable. And it is better than add a boolean variable to indicate which method to use because it need modify too many invoking place and the method is not very clear.

Reply:
Duplicated code is not elegant，I mean you can do like this:

Status Tablet::create_rowset_writer(const Version& version ...
    RowsetWriterContext context = _init_context();
    context.version = version;
    ...
}

Status Tablet::create_rowset_writer(const int64_t& txn_id, const PUniqueId& load_id...
    RowsetWriterContext context = _init_context();
    context.txn_id = txn_id;  
    context.load_id = load_id;
    ...
}

RowsetWriterContext _init_context() {
    RowsetWriterContext context;                                                                    
    context.rowset_id = StorageEngine::instance()->next_rowset_id();                                
    context.tablet_uid = tablet_uid();                                                              
    context.tablet_id = tablet_id();                                                                
    context.partition_id = partition_id();                                                          
    context.tablet_schema_hash = schema_hash();                                                     
    context.data_dir = data_dir();                                                                  
    context.rowset_type = tablet_meta()->preferred_rowset_type();                                   
    // Alpha Rowset will be removed in the future, so that if the tablet's default rowset type is   
    // alpah rowset, then set the newly created rowset to storage engine's default rowset.          
    if (context.rowset_type == ALPHA_ROWSET) {                                                      
        context.rowset_type = StorageEngine::instance()->default_rowset_type();                     
    }                                                                                               
    context.path_desc = tablet_path_desc();                                                         
    context.tablet_schema = &(tablet_schema());                                                     
    context.rowset_state = rowset_state;
    context.segments_overlap = overlap; 
    return context;                                                            
}

Reply:
I see, thanks. I will modify it.